### ISSUE TITLE
Cannot make PWA/SPA work with UP

### ISSUE THREAD

**Issue by trevorseo** [apex/up#651]
## Prerequisites

* [x] I am running the latest version. (`up upgrade`)
* [x] I searched to see if the issue already exists.
* [ ] I inspected the verbose debug output with the `-v, --verbose` flag.
* [x] Are you an Up Pro subscriber?

## Description
I have faced into a problem for PWA/SPA. I wasn't able to make "up" to work with my "dist" directory by adding "type: 'static'" in "up.json" file. So I have created a simple web server with nodejs by myself only to serve "dist" directory. First loading of my web app works perfectly, but the browser(Chrome) downloads "index.html" for all *.js and *.css files instead of its own file if I refresh the page after the first loading. I can fix this behaviour temporarily by removing the cache and loading the page again.

It seems like my web app(PWA/SPA) will work if #637 is added in "Up",,, Am I correct??

### "dist" directory
![image](https://user-images.githubusercontent.com/8579758/39513893-aeb2f9a0-4e30-11e8-9ae3-1c5f0623a750.png)

### "*.js and *.css" downloaded correctly - header
![image](https://user-images.githubusercontent.com/8579758/39515015-c75d8ce2-4e33-11e8-84ff-e3b7b73a4d6a.png)

### "*.js and *.css" downloaded correctly - body
![image](https://user-images.githubusercontent.com/8579758/39514494-45645f32-4e32-11e8-8751-2c18179c766e.png)

### "*.js and *.css" are in problem - header
![image](https://user-images.githubusercontent.com/8579758/39514669-dd22782c-4e32-11e8-9a8f-9670bb763949.png)

### "*.js and *.css" are in problem - body
![image](https://user-images.githubusercontent.com/8579758/39514466-305bcce2-4e32-11e8-98f5-005711c55bf8.png)

### simple web server
``` javascript
const fs = require('fs')
const path = require('path')
const http = require('http')
const { PORT = 3000 } = process.env

// ////////////////////////
// simple logger
// ////////////////////////
function log (level, message, fields = {}) {
  const entry = { level, message, fields }
  console.log(JSON.stringify(entry))
}
// ////////////////////////

// ////////////////////////
// mimeTypes to prepare content-type
// ////////////////////////
const mimeTypes = {
  '.html': 'text/html',
  '.js': 'text/javascript',
  '.css': 'text/css',
  '.json': 'application/json',
  '.png': 'image/png',
  '.jpg': 'image/jpg',
  '.gif': 'image/gif',
  '.wav': 'audio/wav',
  '.mp4': 'video/mp4',
  '.woff': 'application/font-woff',
  '.ttf': 'application/font-ttf',
  '.eot': 'application/vnd.ms-fontobject',
  '.otf': 'application/font-otf',
  '.svg': 'application/image/svg+xml'
}
// ////////////////////////

http.createServer((req, res) => {
  // ////////////////////////
  // 1. prepare url
  // ////////////////////////
  let url = req.url
  log('info', '1. prepare url', { url })
  url = url.split('?')[0]
  log('info', '1. prepare url - remove query', { url })
  // ////////////////////////

  // ////////////////////////
  // 2. prepare filePath
  // ////////////////////////
  let directoryPath = __dirname + '/../dist'
  let indexFilePath = directoryPath + '/index.html'
  let filePath = null
  if (url === '/') {
    filePath = indexFilePath
  } else {
    filePath = directoryPath + req.url
  }
  let fileName = filePath.split('/').reverse()[0]
  log('info', '2. prepare filePath', { fileName, filePath })
  // ////////////////////////

  // ////////////////////////
  // 3. check file existence
  // ////////////////////////
  let isFile = false
  if (fs.existsSync(filePath)) {
    isFile = true
  } else {
    let newFilePath = directoryPath + '/' + fileName
    if (fs.existsSync(newFilePath)) {
      filePath = newFilePath
      isFile = true
    }
  }
  log('info', '3. check file existence', { isFile, filePath })
  // ////////////////////////

  // ////////////////////////
  // 4. handle file existence exception
  // ////////////////////////
  if (!isFile) {
    filePath = indexFilePath
  }
  fileName = filePath.split('/').reverse()[0]
  log('info', '4. handle file existence exception', { fileName, filePath })
  // ////////////////////////

  // ////////////////////////
  // 5. prepare contentType
  // ////////////////////////
  let extname = String(path.extname(filePath)).toLowerCase()
  let contentType = mimeTypes[extname] || 'application/octet-stream'
  log('info', '5. prepare contentType', { fileName, contentType })
  // ////////////////////////

  // ////////////////////////
  // 6. prepare file and send response
  // ////////////////////////
  fs.readFile(filePath, (error, content) => {
    log('info', '6. prepare file and send response', { filePath })
    res.writeHead(200, { 'Content-Type': contentType })
    res.end(content, 'utf-8')
  })
  // ////////////////////////
}).listen(PORT, () => {
  log('info', 'server listening started', { PORT })
})

```
## Steps to Reproduce

Describe the steps required to reproduce the issue if applicable.

**Comment by tj** [comment]
Sounds like more of a service worker configuration if the initial load works fine. 

For the static file serving this should serve `./dist`:

```json
{
  "type": "static",
  "static": { "dir": "dist" }
}
```

The issue you mentioned is for scenarios when you need to try sending the request to your app, and only on a 404 Up will try to find a file for that request, just giving you a bit more control. Currently you can use `static.dir` with dynamic apps as well but it's greedy and tries to find the file before it ever attempts sending the request to your app.

I guess I'd double check that it works fine without the service worker installed, I'm not too familiar with configuring those yet so I'm not a huge help there :D

**Comment by trevorseo** [comment]
I was so focusing on making this works, so I forgot the reason why I ended up making the "simple web server" at the time I reported this issue.

### SPA and PWA works with the UP
After I added below in ```up.json```, my SPA and PWA works perfectly.
```json
  "type": "static",
  "static": { "dir": "dist" },
  "redirects": {
    "/*": {
      "location": "/",
      "status": 200
    }
  },
  "stages": {
    "staging": {
      "domain": "staging-yourapp.domain.com"
    },
    "production": {
      "domain": "yourapp.domain.com"
    }
  }
```

But it works only using the **custom domain name** not with **auto generated urls** like below.
- https://xxx.execute-api.us-east-2.amazonaws.com/staging/
- https://xxx.execute-api.us-east-2.amazonaws.com/production/

### Findings
- It is easier to use ```static.dir``` option when working with SPA and PWA than adding custom web server.
- PWA is safe with the UP since service-worker.js works fine.
- For SPA, it works only using the **custom domain name** not with **auto generated urls**. My browser gives out "403 error" because it tries to get css, js files from "/" not from "staging/" or production/". (this was the reason why I made the **simple web server** for myself.
![image](https://user-images.githubusercontent.com/8579758/39613391-ff7ae948-4fa1-11e8-9e3e-3f4a9171be95.png)

**Comment by tj** [comment]
Ahhh yeah the asset serving from a prefix becomes a pain, some frameworks allow specifying some kind of base path, but it's annoying either way.

I'll close this one for now, unfortunately there's no way to remove that `/stage` prefix from API Gateway at the moment. I like to chuck a real domain on there soon too, definitely helps with dev

**Comment by nodkrot** [comment]
Hey @tj, so the only solution for this now is using custom domain?

**Comment by tj** [comment]
@nodkrot yeah if you plan on using a domain longer-term anyway I would definitely suggest using `stage.myapp.com` or similar early on. For now API Gateway still requires the prefixing

**Comment by nodkrot** [comment]
Got it, thx. Will maybe i can do something with env variables and manual redirect logic
